### PR TITLE
Try MIR inlining leaf functions by default

### PR DESCRIPTION
**Pull request by tmiasko**
Based on #77307 and #68828.

### CODE REVIEW

File: compiler/rustc_mir/src/transform/inline.rs [hunk 1]
```diff
@@ -885,3 +903,48 @@ impl<'a, 'tcx> MutVisitor<'tcx> for Integrator<'a, 'tcx> {
...
+    }
+
+    use rustc_hir::def::DefKind;
+    if !matches!(tcx.def_kind(did), DefKind::Fn | DefKind::AssocFn) {
+        debug!("is_trivial_mir = false (not a function)");
+        // Only inline functions, don't look at constants here.
+        return false;
+    }
+
+    if !did.is_local() {
```
Comment (lcnr) [inline compiler/rustc_mir/src/transform/inline.rs:921]: this is unreachable. In decoding we just store `false` in this case :laughing:

**Review by lcnr — COMMENTED** [review]
so inlining parts `iter::Empty` fixes the `deeply-nested` test, gj on figuring all of this out.

I don't know enough about about the other touched mir opts or why they are necessary as part of this PR, so
r? @oli-obk

**Comment by the8472** [comment]
> Finished benchmarking try commit ([032277c](https://github.com/rust-lang/rust/commit/032277c8bdf5002b9a9d01a6a1c5db9ea04fc839)): [comparison url](https://perf.rust-lang.org/compare.html?start=410a546fc593272738edf53fdca020f89595752b&end=032277c8bdf5002b9a9d01a6a1c5db9ea04fc839).


The max-rss results look substantially different (worse) than the noise run, is this expected?

**Review by oli-obk — COMMENTED** [review]
afaict everything after f0f5e19 is not actually necessary for this PR, right? 

> The max-rss results look substantially different (worse) than the noise run, is this expected?

I would expect inlining to cause an increase in memory usage as MIR bodies now contain duplicates of other MIR bodies.

File: compiler/rustc_mir/src/transform/inline.rs [hunk 2]
```diff
@@ -885,3 +903,48 @@ impl<'a, 'tcx> MutVisitor<'tcx> for Integrator<'a, 'tcx> {
         }
     }
 }
+
+pub fn is_trivial_mir(tcx: TyCtxt<'tcx>, did: DefId) -> bool {
```
Comment (nagisa) [inline compiler/rustc_mir/src/transform/inline.rs:907]: This  could really use a more semantic name. A "trivial" can mean many things, and what it means in this context is not really specified.

Given that its purpose is to check that the body is trivially inlineable, a name for your consideration: `is_trivially_inlineable` or perhaps `is_leaf_mir`.

File: compiler/rustc_mir/src/transform/simplify.rs [hunk 3]
```diff
@@ -375,15 +376,19 @@ struct UsedLocals {
...
 
 impl UsedLocals {
     /// Determines which locals are used & unused in the given body.
-    fn new(body: &Body<'_>) -> Self {
+    fn new(tcx: TyCtxt<'_>, body: &Body<'_>) -> Self {
         let mut this = Self {
             increment: true,
             arg_count: body.arg_count.try_into().unwrap(),
             use_count: IndexVec::from_elem(0, &body.local_decls),
+            preserve_var_debug_info: tcx.sess.opts.optimize != OptLevel::Aggressive,
```
Comment (nagisa) [inline compiler/rustc_mir/src/transform/simplify.rs:391]: What criteria was used in choosing this? Other compilers generally don't try to preserve any debuginfo after `-O2` which is `OptLevel::Default`. Futhermore, this is a MIR optimisation so it is unclear to me that deriving this from `-O` flag is actually what we want?

**Comment by tmiasko** [comment]
The unicode_normalization-check full is now at -0.4%, fixing the previous +16.4% regression.

The results for packed-simd-debug & packed-simd-opt show a large regression. That is one things I would like to investigate more. The first impression is that this crate has a lot of trivial inlining opportunities while at the same time doing almost not code generation, so we are paying the cost of inlining but receive no benefits.

> afaict everything after f0f5e19 is not actually necessary for this PR, right?

One additional change that follows is an extended removal of dead locals. This is something I would like to do at some point (not necessarily in the form here). The current implementation of SimplifyLocals is very conservative, and retains locals from removed dead code when they are referenced by debuginfo (e.g., it retains locals from `debug_assert_eq!` when debug assertions are disabled).

**Pull request closed without merging**